### PR TITLE
CASSANDRA-18500: Add guardrail for partition size

### PR DESCRIPTION
**Pull request by adelapena**
The patch adds a new guardrail for partition size that is checked when writing a sstable (flush, compaction, etc.)

Triggering the failure threshold will emit an ERROR log message and a diagnostic event, but it won't abort the sstable write operation.

This guardrail replaces the `compaction_large_partition_warning_threshold` property, which is now marked as deprecated. Both checks will exist until we remove `compaction_large_partition_warning_threshold` on the next major. The reason for keeping both properties and not just renaming `compaction_large_partition_warning_threshold` to `partition_size_warn_threshold` is that the behaviour, log messages, etc. are slightly different.

It's worth mentioning that the existing `compaction_large_partition_warning_threshold`:
* Isn't actually triggered only on compaction, but on any sstable write.
* Cannot be dynamically updated.
* Doesn't emit any kind of diagnostic events.

### CODE REVIEW

File: conf/cassandra.yaml [hunk 1]
```diff
@@ -1804,6 +1805,15 @@ drop_compact_storage_enabled: false
 # write_consistency_levels_warned: []
 # write_consistency_levels_disallowed: []
 #
+# Guardrail to warn or fail when writing partitions larger than threshold.
+# The guardrail is only checked when writing sstables (flush, compaction, streaming, etc.), and exceeding the fail
+# threshold on that moment will only log an error message, without interrupting the operation.
+# This operates on a per-sstable basis, so it won't detect a large partition if it is spread across multiple sstables.
+# The warning threshold replaces the deprecated config property compaction_large_partition_warning_threshold.
+# The two thresholds default to null to disable.
```
Comment (bereng) [inline conf/cassandra.yaml:1822]: Can you add a few valid examples to see syntax/units? 100MiB, 100MB, 10KB,...

File: src/java/org/apache/cassandra/db/guardrails/GuardrailsMBean.java [hunk 2]
```diff
@@ -468,6 +468,30 @@
...
 
+    /**
+     * @return The threshold to warn when encountering partitions larger than threshold, as a string  formatted as in,
+     * for example, {@code 10GiB}, {@code 20MiB}, {@code 30KiB} or {@code 40B}. A {@code null} value means disabled.
+     */
+    @Nullable
+    String getPartitionSizeWarnThreshold();
+
+    /**
+     * @return The threshold to fail when encountering partitions larger than threshold, as a string  formatted as in,
```
Comment (bereng) [inline src/java/org/apache/cassandra/db/guardrails/GuardrailsMBean.java]: I would reword all the 'fail' variant comments to highlight the operation won't fail and it will only get logged at ERROR level. Otherwise it's confusing imo.

File: test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailPartitionSizeTest.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import static java.nio.ByteBuffer.allocate;
+
+/**
+ * Tests the guardrail for the size of partition size, {@link Guardrails#partitionSize}.
+ * <p>
+ * This test only includes the activation of the guardrail during sstable writes, focusing on the emmitted log messages.
+ * The tests for config, client warnings and diagnostic events are in
+ * {@link org.apache.cassandra.db.guardrails.GuardrailPartitionSizeTest}.
+ */
+public class GuardrailPartitionSizeTest extends GuardrailTester
```
Comment (bereng) [inline test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailPartitionSizeTest.java:40]: Do you test anywhere the thresholds can be dynamically changed?

File: conf/cassandra.yaml [hunk 4]
```diff
@@ -1541,7 +1541,8 @@ batch_size_fail_threshold: 50KiB
 # Log WARN on any batches not of type LOGGED than span across more partitions than this limit
 unlogged_batch_across_partitions_warn_threshold: 10
 
-# Log a warning when compacting partitions larger than this value
+# Log a warning when compacting partitions larger than this value.
+# As of Cassandra 5.0, this property is deprecated in favour of partition_size_warn_threshold.
```
Comment (Maxwell-Guo) [inline conf/cassandra.yaml:1545]: can change to "5.0 and later versions ", I think this will be more  rigorous

File: conf/cassandra.yaml (hunk 1 above)
Comment (Maxwell-Guo) [reply conf/cassandra.yaml:1822]: My question is : should we regardless of the stream situation? Because the stream data is generated by flush and compaction or logging is enough ?

File: src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java [hunk 5]
```diff
@@ -324,6 +325,19 @@ protected FileHandle openDataFile(long lengthOverride, StatsMetadata statsMetada
         return dataFile;
     }
 
+    private void guardPartitionSize(DecoratedKey key, long rowSize)
+    {
+        if (Guardrails.partitionSize.triggersOn(rowSize, null))
+        {
+            String what = String.format("%s.%s:%s on sstable %s",
```
Comment (Maxwell-Guo) [inline src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java:332]: our experience is we prefer metadata.keyspace +    metadata.name +                              metadata().partitionKeyType.getString(key.getKey()) +   getFilename() to contact these strings than String.format. 
as String.format  will cost more time ,But it seems cassandra just use String.format  all time.So this comment can be ignored.

File: test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailPartitionSizeTest.java (hunk 3 above)
Comment (adelapena) [reply test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailPartitionSizeTest.java:40]: That is tested on the utest, whereas this dtest is just focused on verifying the not-redacted messages. But I have nevertheless added a new `testPartitionSizeWithDynamicUpdate`.

File: src/java/org/apache/cassandra/db/guardrails/GuardrailsMBean.java (hunk 2 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/db/guardrails/GuardrailsMBean.java]: I guess the meaning of "fail" on all background guardrails like this one is logging an error and emit a failure event. Not very sure how to rename, but I have added a line telling that it will not throw an exception to avoid stopping the write.

File: conf/cassandra.yaml (hunk 4 above)
Comment (adelapena) [reply conf/cassandra.yaml:1545]: I think that "as of" means "starting from", or versions >= 5.0, so the "and later versions" part seems redundant. Also, since it's a deprecation I think it wouldn't make sense for it to not be deprecated in later versions.

File: src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java (hunk 5 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java:332]: Indeed `String.format` is everywhere. In this case, the message is only built when the guardrail is actually triggered, which shouldn't be very often.

File: conf/cassandra.yaml (hunk 1 above)
Comment (adelapena) [reply conf/cassandra.yaml:1822]: > Can you add a few valid examples to see syntax/units? 100MiB, 100MB, 10KB,...

Yes, I have added a couple of examples about the expected values, like the ones on the JMX methods.

> My question is : should we regardless of the stream situation? Because the stream data is generated by flush and compaction or logging is enough ?

I'm not sure I understand what we should do and what you mean by the stream situation. Tables streamed from other nodes would have triggered the guardrail in their node of origin. We probably don't want to scan them again in the receiver node as they arrive to avoid the performance cost of scanning them. Also, large partitions in a streamed table can be later be detected by the guardrail when those partitions are compacted.

File: conf/cassandra.yaml (hunk 4 above)
Comment (Maxwell-Guo) [reply conf/cassandra.yaml:1545]: ok

File: src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java (hunk 5 above)
Comment (Maxwell-Guo) [reply src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java:332]: yes ,I found C* use String.format almost everywhere. So I said this comment can be ignored.

File: conf/cassandra.yaml (hunk 1 above)
Comment (Maxwell-Guo) [reply conf/cassandra.yaml:1822]: These sstables that are going to stream to other nodes are generated by flush or compaction , and we may just need to use the guardrail to detecte at flush or compaction stage. 

The reveiver  or the sending node do not need to use the guardrail to detecte at stream stage. 

That is what I mean.

Comment (bereng) [reply conf/cassandra.yaml:1822]: Thx on the examples.

File: test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailPartitionSizeTest.java (hunk 3 above)
Comment (bereng) [reply test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailPartitionSizeTest.java:40]: +1

File: src/java/org/apache/cassandra/db/guardrails/GuardrailsMBean.java (hunk 2 above)
Comment (bereng) [reply src/java/org/apache/cassandra/db/guardrails/GuardrailsMBean.java]: +1

**Comment by bereng** [comment]
I don't have further comments so far besides that we're missing a final CI run

File: conf/cassandra.yaml (hunk 1 above)
Comment (adelapena) [reply conf/cassandra.yaml:1822]: > These sstables that are going to stream to other nodes are generated by flush or compaction , and we may just need to use the guardrail to detecte at flush or compaction stage.

I see, I have updated the comment to exclude streaming.

**Comment by adelapena** [comment]
CI is running here:
* https://app.circleci.com/pipelines/github/adelapena/cassandra/2906/workflows/e322a36b-b38a-4387-9464-4eeede1edbd9
* https://app.circleci.com/pipelines/github/adelapena/cassandra/2906/workflows/c38b9241-4dd0-4180-b7a8-b006ba52c8cc

**Comment by Maxwell-Guo** [comment]
+1 on this patch

**Comment by bereng** [comment]
It seems circle is having issues and the CI runs don't look too nice. Could we re-trigger them please?

**Comment by adelapena** [comment]
Yes, I've already done it:
* https://app.circleci.com/pipelines/github/adelapena/cassandra/2906/workflows/9ce9fb2a-fe2a-4619-ba1c-ce4d29a1016c
* https://app.circleci.com/pipelines/github/adelapena/cassandra/2906/workflows/d28b205c-1b51-49aa-959a-948be3fa06eb

All green except for a consistent failure on `GetFullQueryLogTest` that has just been introduced in trunk by [CASSANDRA-18550](https://issues.apache.org/jira/browse/CASSANDRA-18550)

**Comment by bereng** [comment]
+1
